cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -247,4 +269,34 @@ public WorkflowToken deepCopy() {
     }
     return builder.build();
   }
+
+  private void ensureEnoughMemoryAvailable(String key, Value value) {
+    int left = bytesLeft;
+    left -= key.length();
+    if (left <= 0) {
+      throw new IllegalStateException(String.format("Exceeded maximum permitted size of workflow token '%sMB' while " +
+                                                      "adding key '%s'. Current size is '%sMB'. Please increase the " +
+                                                      "maximum permitted size by setting the parameter '%s' in " +
+                                                      "cdap-site.xml to add more values.",
+                                                    maxSizeBytes / (1024 * 1024), key,
+                                                    currentSizeBytes / (1024 * 1024),
+                                                    Constants.AppFabric.WORKFLOW_TOKEN_MAX_SIZE_MB));
+    }
+    left -= value.toString().length();
+    if (left <= 0) {
+      throw new IllegalStateException(String.format("Exceeded maximum permitted size of workflow token '%sMB' while " +
```

Comment:
You can further simplify the logic into one throw:

```java
/**
 * Appends the given key/value to the given map.
 *
 * @throws IllegalArgumentException if the max size will be exceeded after adding the given key/value.
 */
void append(String key, Value value, Map<String, List<NodeValue>> nodeValues) {
  int left = bytesLeft - key.length;
  left = (left > 0) ? left - value.length : left;
  if (left <= 0) {
    throw ...
  }
  nodeValues.get(key).add(new NodeValue(nodeName, value));
  bytesLeft = left;
}
```

And I don't see why this method needs further breakdown (it's less than 10 lines).